File: src/utils/hashing.ts
import * as crypto from "crypto";
import * as fs from "fs/promises";
import * as path from "path";

/**
 * Compute SHA256 hash of a file
 */
export async function computeFileHash(filePath: string): Promise<string> {
    const buffer = await fs.readFile(filePath);
    return crypto.createHash("sha256").update(buffer).digest("hex");
}

/**
 * Compute SHA256 hash of a directory based on the design spec:
 * SHA256(concat(sorted list of: relative_path + ":" + file_hash))
 */
export async function computeDirectoryHash(dirPath: string): Promise<string> {
    const files = await getAllFiles(dirPath);
    const hashes: { relPath: string; hash: string }[] = [];

    for (const file of files) {
        const relPath = path.relative(dirPath, file).replace(/\\/g, "/");
        // Exclude hidden files
        if (relPath.split("/").some(part => part.startsWith("."))) {
            continue;
        }
        const hash = await computeFileHash(file);
        hashes.push({ relPath, hash });
    }

    // Sort lexicographically by relative_path
    hashes.sort((a, b) => a.relPath.localeCompare(b.relPath));

    const concatString = hashes.map(h => `${h.relPath}:${h.hash}`).join("");
    return crypto.createHash("sha256").update(concatString).digest("hex");
}

async function getAllFiles(dir: string): Promise<string[]> {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    const files = await Promise.all(
        entries.map((entry) => {
            const res = path.resolve(dir, entry.name);
            return entry.isDirectory() ? getAllFiles(res) : [res];
        })
    );
    return files.flat();
}

File: vhl_protocol/utils/hashing.py
import hashlib
import os

def compute_file_hash(file_path: str) -> str:
    """Compute SHA256 hash of a file."""
    sha256_hash = hashlib.sha256()
    with open(file_path, "rb") as f:
        # Read and update hash string value in blocks of 4K
        for byte_block in iter(lambda: f.read(4096), b""):
            sha256_hash.update(byte_block)
    return sha256_hash.hexdigest()

def compute_directory_hash(dir_path: str) -> str:
    """
    Compute SHA256 hash of a directory based on the design spec:
    SHA256(concat(sorted list of: relative_path + ":" + file_hash))
    """
    hashes = []
    
    for root, dirs, files in os.walk(dir_path):
        # Exclude hidden directories
        dirs[:] = [d for d in dirs if not d.startswith('.')]
        
        for file in files:
            # Exclude hidden files
            if file.startswith('.'):
                continue
                
            abs_path = os.path.join(root, file)
            rel_path = os.path.relpath(abs_path, dir_path).replace("\\", "/")
            
            file_hash = compute_file_hash(abs_path)
            hashes.append((rel_path, file_hash))
            
    # Sort lexicographically by relative_path
    hashes.sort(key=lambda x: x[0])
    
    concat_string = "".join(f"{rel_path}:{file_hash}" for rel_path, file_hash in hashes)
    return hashlib.sha256(concat_string.encode("utf-8")).hexdigest()
